third_party/lua/luadebug.lua

```diff
@@ -162,41 +161,188 @@ local function format_stack_frame_info(info)
     return format_loc(source, info.currentline) .. " in " .. namewhat .. " " .. name
 end
 
-local normalized_paths = {}
+local suffix_arrays = {}
+
+local function normalize_file(file)
+    return file and (file:gsub('^@', ''):gsub('\n', '')) or ''
+end
 
 -- This function is on a hot path of a debugger hook.
 -- Try very hard to avoid wasting of CPU cycles, so reuse
 -- previously calculated results via memoization
-local function normalize_path(file)
+local function build_suffix_array(file)
     local decorated_name = file
-    if normalized_paths[decorated_name] ~= nil then
-        return normalized_paths[decorated_name]
+    if suffix_arrays[decorated_name] ~= nil then
+        return suffix_arrays[decorated_name]
+    end
+
+    local suffixes = {}
+    file = normalize_file(file)
+    --[[
+        we would very much prefer to use simple:
+           for v in file:gmatch('[^/]+') do
+        loop here, but string.gmatch is very slow
+        but we actually need several strchr's here
+    ]]
+    local i = 0
+    local j
+    local v
+    repeat
+        j = file:find('/', i, true)
+        if j ~= nil then
+            v = file:sub(i, j - 1)
+            if v ~= '.' then -- just skip '.' for current dir
```

Comment:
> '.' is not always used just for current dir, it can be a mandatory part of a path, like '../'

in the case of `if v ~= '.' then ` statement, '.' was exactly about the case `xxx/.yyy` case. And it adds nothing to the suffix tree.

When you were adding breakpoint for `../build/one/test.lua:2` you see it properly visualized (that's expected as code_listing was working in the context of current directory at the start of invocation), it then created suffix tree in a form `test.lua/one/build/..` which was never hit.

If you would add it in a form `one/test.lua:2` (partial path as you were shown in gdb examples before)  then it would be working.

Reply:
My point is not about this exact line, but overall usability. It is not possible to create a breakpoint by full path nor by relative path, that goes one directory up. This seems like a major limitation.

```
luadebug.lua: Loaded for 2.11.0-entrypoint-999-g0241645fd
break via dbg() => ./main.lua:8 in chunk at ./main.lua:0
   5   
   6   dbg()
   7   
   8=> hello_1()
   9   hello_2()
  10   
  11   local compat = require('tarantool').compat
luadebug.lua> b /home/ubuntu/Code/review/tarantool-tdbg/build/main.lua:11
DEBUG: before: 	/home/ubuntu/Code/review/tarantool-tdbg/build/main.lua
DEBUG: after: 	/home/ubuntu/Code/review/tarantool-tdbg/build/main.lua
Added breakpoint /home/ubuntu/Code/review/tarantool-tdbg/build/main.lua:11.
   8   hello_1()
   9   hello_2()
  10   
DEBUG: before: 	/home/ubuntu/Code/review/tarantool-tdbg/build/main.lua
  11=>●local compat = require('tarantool').compat
  12   
  13   print(compat.dump())
luadebug.lua> c
hello 1!
hello 2!
DEBUG: before: 	@./main.lua
DEBUG: after: 	./main.lua
DEBUG: before: 	@./main.lua
require('tarantool').compat({
    yaml_pretty_multiline            = 'default',
    json_escape_forward_slash        = 'default',
    fiber_channel_close_mode         = 'default',
    box_cfg_replication_sync_timeout = 'default',
})
```

Reply:
`break ../c/a/b.lua` -> `abspath(getcwd() + "../c/a/b.lua")` (at the moment of setting breakpoint)
will not work, if debug.getinfo(n, "S").source contains relative path.

`break ./c/a/b.lua` -> `abspath(getcwd() + "c/a/b.lua")` (at the moment of setting breakpoint)
will not work, if debug.getinfo(n, "S").source contains relative path.

`..` or `.` inside path should be simplified:

`break x/../c/a/b.lua` -> `c/a/b.lua`
`break x/./c/a/b.lua` -> `x/c/a/b.lua`

Reply:
Here is majority of introduced changed, as per @Mons specification above:

```diff
--- a/third_party/lua/luadebug.lua
+++ b/third_party/lua/luadebug.lua
@@ -161,28 +161,51 @@ local function format_stack_frame_info(info)
     return format_loc(source, info.currentline) .. " in " .. namewhat .. " " .. name
 end
 
-local suffix_arrays = {}
+local fio = require 'fio'
+local cwd = fio.cwd()
 
-local function normalize_file(file)
-    return file and (file:gsub('^@', ''):gsub('\n', '')) or ''
+local suffix_arrays = {} -- memoize file x tree resuls
+local normalized_files = {} -- memoize normalization results
+
+--[[
+    Normalize `debug.getinfo().source` names:
+    - remove leading '@';
+    - remove trailing new lines (if any);
+    - if there is leading '.' or '..' then calculate full path
+      using current directory.
+]]
+local function normalize_file(file, resolve_dots)
+    local srcfile = file
+    if not file then
+        return ''
+    end
+    if normalized_files[file] then
+        return normalized_files[file]
+    end
+    file = file:gsub('^@', ''):gsub('\n', '')
+    if resolve_dots and file:sub(1,1) == '.' then
+        file = fio.abspath(fio.pathjoin(cwd, file))
+    end
+    normalized_files[srcfile] = file
+    return file
 end
 
 -- This function is on a hot path of a debugger hook.
 -- Try very hard to avoid wasting of CPU cycles, so reuse
 -- previously calculated results via memoization
-local function build_suffix_array(file)
+local function build_suffix_array(file, resolve_dots)
     local decorated_name = file
     if suffix_arrays[decorated_name] ~= nil then
         return suffix_arrays[decorated_name]
     end
 
     local suffixes = {}
-    file = normalize_file(file)
+    file = normalize_file(file, resolve_dots)
     --[[
         we would very much prefer to use simple:
            for v in file:gmatch('[^/]+') do
         loop here, but string.gmatch is very slow
-        but we actually need several strchr's here
+        and that we actually need here are simple strchr's.
     ]]
     local i = 0
     local j
@@ -192,7 +215,11 @@ local function build_suffix_array(file)
         if j ~= nil then
             v = file:sub(i, j - 1)
             if v ~= '.' then -- just skip '.' for current dir
-                table.insert(suffixes, 1, v)
+                if v == '..' then
+                    table.remove(suffixes)
+                else
+                    table.insert(suffixes, 1, v)
+                end
             end
             i = j + 1
         end
@@ -200,7 +227,11 @@ local function build_suffix_array(file)
     -- don't forget to process the trailing segment
     v = file:sub(i, #file)
     if v ~= '.' then -- just skip '.' for current dir
-        table.insert(suffixes, 1, v)
+        if v == '..' then
+            table.remove(suffixes)
+        else
+            table.insert(suffixes, 1, v)
+        end
     end
 
     suffix_arrays[decorated_name] = suffixes
@@ -269,7 +295,7 @@ end
 
 --[[
     Given suffix tree T try to remove suffix array A
-    from the tree. Laaf and intermediate nodes will be
+    from the tree. Leafs and intermediate nodes will be
     cleaned up only once their reference counter '$ref'
     will reach 1.
 ]]
@@ -319,8 +345,8 @@ local function has_active_breakpoints()
 end
 
 local function add_breakpoint(file, line)
-    local suffixes = build_suffix_array(file)
-    local normfile = normalize_file(file)
+    local suffixes = build_suffix_array(file, true)
+    local normfile = normalize_file(file, true)
     -- save direct hash (for faster lookup): line -> file
     if not breakpoints[line] then
         breakpoints[line] = {}
@@ -334,11 +360,11 @@ local function add_breakpoint(file, line)
 end
 
 local function remove_breakpoint(file, line)
-    local normfile = normalize_file(file)
+    local normfile = normalize_file(file, true)
     if breakpoints[line] == nil then
         return false
     end
-    local suffixes = build_suffix_array(file)
+    local suffixes = build_suffix_array(file, true)
     remove_suffix_array(breakpoints[line], suffixes)
     breakpoints[normfile][line] = nil
 end
```

Reply:
Having now full path saved, we need to simplify breakpint removal scenario then (as entering full path there, to remove multiple breakpoints, would be very, very anoying). I plan to use an approach similar to `gdb`, when we pass the number for deleting a numbered breakpoint from the list.